File: scripts/example2/dev/networkNode.js
const express = require('express')
const app = express()
const bodyParser = require('body-parser');
const Blockchain = require('./blockchain');

// this library does for us is it creates a unique string, a unique random string for us. And we're going to use that string as this network nodes address.
const uuid = require('uuid');
// const { json } = require('body-parser');
const nodeAddress = uuid.v1().split('-').join('');

// So in order to get access to this variable in our network, a node file, we are using this process,
const port = process.argv[2];

const rp = require('request-promise');

// we want to make an instance of our block chain so we will say consed bitcoin equals a new block
const bitcoin = new Blockchain();

// So basically, all these two lines are doing is they're saying if a request comes in with JSON data or with form data, we simply want to pass that data so that we can access it in any of these routes.
// So any route we hit our data is first going to go through the body parser so we can access the data and then we can use the data in whichever endpoint we'll be receiving it.
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

// We can handle different end points or different routes.
// For example, we have a get endpoint, which is just slash. And with this endpoint, we are sending back the response of Hello World and then this whole server is listening on Port 3000

// we can test to make sure that this end point worked correctly, as we can check out our entire block chain by hitting our block chain and point
app.get('/blockchain', function (req, res) {
  res.send(bitcoin);
});

// app.post('/transaction', function (req, res) {
//     // console.log(req.body);
//     // res.send(`The amount of the transaction is ${req.body.amount} bitcoin.`);
//     const blockIndex = bitcoin.createNewTransaction(req.body.amount, req.body.sender, req.body.recipient);
//     res.json({ note: `Transaction will be added in block ${blockIndex}.`});
// });


// create a new transaction
app.post('/transaction', function(req, res) {
	const newTransaction = req.body;
	const blockIndex = bitcoin.addTransactionToPendingTransactions(newTransaction);
	res.json({ note: `Transaction will be added in block ${blockIndex}.` });
});

// broadcast transaction
app.post('/transaction/broadcast', function(req, res) {
	const newTransaction = bitcoin.createNewTransaction(req.body.amount, req.body.sender, req.body.recipient);
	bitcoin.addTransactionToPendingTransactions(newTransaction);

	const requestPromises = [];
	bitcoin.networkNodes.forEach(networkNodeUrl => {
		const requestOptions = {
			uri: networkNodeUrl + '/transaction',
			method: 'POST',
			body: newTransaction,
			json: true
		};

		requestPromises.push(rp(requestOptions));
	});

	Promise.all(requestPromises)
	.then(data => {
		res.json({ note: 'Transaction created and broadcast successfully.' });
	});
});

// this mine endpoint will do is it will mine a new block for us or create a new block for us,
// mine a block
app.get('/mine', function(req, res) {
	const lastBlock = bitcoin.getLastBlock();
	const previousBlockHash = lastBlock['hash'];
	const currentBlockData = {
		transactions: bitcoin.pendingTransactions,
		index: lastBlock['index'] + 1
	};
	const nonce = bitcoin.proofOfWork(previousBlockHash, currentBlockData);
	const blockHash = bitcoin.hashBlock(previousBlockHash, currentBlockData, nonce);
	const newBlock = bitcoin.createNewBlock(nonce, previousBlockHash, blockHash);

	const requestPromises = [];
	bitcoin.networkNodes.forEach(networkNodeUrl => {
		const requestOptions = {
			uri: networkNodeUrl + '/receive-new-block',
			method: 'POST',
			body: { newBlock: newBlock },
			json: true
		};

		requestPromises.push(rp(requestOptions));
	});

	Promise.all(requestPromises)
	.then(data => {
		const requestOptions = {
			uri: bitcoin.currentNodeUrl + '/transaction/broadcast',
			method: 'POST',
			body: {
				amount: 12.5,
				sender: "00",
				recipient: nodeAddress
			},
			json: true
		};

		return rp(requestOptions);
	})
	.then(data => {
		res.json({
			note: "New block mined & broadcast successfully",
			block: newBlock
		});
	});
});


// receive new block
app.post('/receive-new-block', function(req, res) {
	const newBlock = req.body.newBlock;
	const lastBlock = bitcoin.getLastBlock();
	const correctHash = lastBlock.hash === newBlock.previousBlockHash; 
	const correctIndex = lastBlock['index'] + 1 === newBlock['index'];

	if (correctHash && correctIndex) {
		bitcoin.chain.push(newBlock);
		bitcoin.pendingTransactions = [];
		res.json({
			note: 'New block received and accepted.',
			newBlock: newBlock
		});
	} else {
		res.json({
			note: 'New block rejected.',
			newBlock: newBlock
		});
	}
});

// So now in our block chain, we want to be able to create a network and we want to have a way of registering all these different nodes with our network.
// So right now, we're going to make a couple more end points that will make registering nodes with our network possible.
// register a node and broadcast it the network
app.post('/register-and-broadcast-node', function(req,res) {
    const newNodeUrl = req.body.newNodeUrl;
	if (bitcoin.networkNodes.indexOf(newNodeUrl) == -1) bitcoin.networkNodes.push(newNodeUrl);

	const regNodesPromises = [];
	bitcoin.networkNodes.forEach(networkNodeUrl => {
		const requestOptions = {
			uri: networkNodeUrl + '/register-node',
			method: 'POST',
			body: { newNodeUrl: newNodeUrl },
			json: true
		};

		regNodesPromises.push(rp(requestOptions));
	});

	Promise.all(regNodesPromises)
	.then(data => {
		const bulkRegisterOptions = {
			uri: newNodeUrl + '/register-nodes-bulk',
			method: 'POST',
			body: { allNetworkNodes: [ ...bitcoin.networkNodes, bitcoin.currentNodeUrl ] },
			json: true
		};

		return rp(bulkRegisterOptions);
	})
	.then(data => {
		res.json({ note: 'New node registered with network successfully.' });
	});
});

// register a node with the network
app.post('/register-node', function(req,res) {
    const newNodeUrl = req.body.newNodeUrl;
	const nodeNotAlreadyPresent = bitcoin.networkNodes.indexOf(newNodeUrl) == -1;
	const notCurrentNode = bitcoin.currentNodeUrl !== newNodeUrl;
	if (nodeNotAlreadyPresent && notCurrentNode) bitcoin.networkNodes.push(newNodeUrl);
	res.json({ note: 'New node registered successfully with node.' });
});

// register multiple nodes at once
app.post('/register-nodes-bulk', function(req, res) {
	const allNetworkNodes = req.body.allNetworkNodes;
	allNetworkNodes.forEach(networkNodeUrl => {
		const nodeNotAlreadyPresent = bitcoin.networkNodes.indexOf(networkNodeUrl) == -1;
		const notCurrentNode = bitcoin.currentNodeUrl !== networkNodeUrl;
		if (nodeNotAlreadyPresent && notCurrentNode) bitcoin.networkNodes.push(networkNodeUrl);
	});

	res.json({ note: 'Bulk registration successful.' });
});

// consensus
app.get('/consensus', function(req, res) {
	const requestPromises = [];
	bitcoin.networkNodes.forEach(networkNodeUrl => {
		const requestOptions = {
			uri: networkNodeUrl + '/blockchain',
			method: 'GET',
			json: true
		};

		requestPromises.push(rp(requestOptions));
	});

	Promise.all(requestPromises)
	.then(blockchains => {
		const currentChainLength = bitcoin.chain.length;
		let maxChainLength = currentChainLength;
		let newLongestChain = null;
		let newPendingTransactions = null;

		blockchains.forEach(blockchain => {
			if (blockchain.chain.length > maxChainLength) {
				maxChainLength = blockchain.chain.length;
				newLongestChain = blockchain.chain;
				newPendingTransactions = blockchain.pendingTransactions;
			};
		});


		if (!newLongestChain || (newLongestChain && !bitcoin.chainIsValid(newLongestChain))) {
			res.json({
				note: 'Current chain has not been replaced.',
				chain: bitcoin.chain
			});
		}
		else {
			bitcoin.chain = newLongestChain;
			bitcoin.pendingTransactions = newPendingTransactions;
			res.json({
				note: 'This chain has been replaced.',
				chain: bitcoin.chain
			});
		}
	});
});

// get block by blockHash
app.get('/block/:blockHash', function(req, res) { 
	// So the first thing that we want to do in this end point is we want to use the black hash that is sent in and this request and we can access this black hash and the req programs object.
	const blockHash = req.params.blockHash;
	// use our new get Block method inside of our block block hash and point to retrieve a specific block by its block hash.
	const correctBlock = bitcoin.getBlock(blockHash);
	// All we have to do now is send back this variable or this correct black as a response.
	res.json({
		block: correctBlock
	});
});

// get transaction by transactionId
app.get('/transaction/:transactionId', function(req, res) {
	const transactionId = req.params.transactionId;
	const trasactionData = bitcoin.getTransaction(transactionId);
	res.json({
		transaction: trasactionData.transaction,
		block: trasactionData.block
	});
});

// get address by address
app.get('/address/:address', function(req, res) {
	const address = req.params.address;
	const addressData = bitcoin.getAddressData(address);
	res.json({
		addressData: addressData
	});
});

app.listen(port, function(){
    // the reason that we do this is just so that when our port is actually running, we will know that because we'll see this text.
    console.log(`Listening on port ${port}...`);
});

File: supply-chain-financing_0/Blockchain-Python/networkNode.py
from flask import Flask,request, jsonify,render_template
from blockchain import Blockchain
import json
import requests
import argparse

parser = argparse.ArgumentParser(description='Run flask program')
parser.add_argument(
    '--port',
    action='store',
    dest='port',
    type=int,
    default=8888,
    metavar='PORTNUM',
    help='set port number')
args = parser.parse_args()
app = Flask(__name__,template_folder='templates')
bc = Blockchain(args.port)

@app.route('/blockchain',methods=['GET'])
def blockchain():
    # print(bc.__dict__)
    return json.dumps(bc.__dict__)

@app.route("/transaction",methods=['POST'])
def transaction():
    newTransaction = request.json
    # print(request.data)
    blockIndex = bc.addTransactionToPendingTransaction(newTransaction)
    return {"note":f"Transaction will be added in block {blockIndex}"}

@app.route("/transaction/broadcast",methods=['POST'])
def transactionBroadcast():
    data = request.json
    # newTransaction = bc.createNewTransaction(data['amount'],data['sender'],data['recipient'])
    newTransaction = bc.createNewTransaction(**data)
    bc.addTransactionToPendingTransaction(newTransaction)
    responseAll = []
    for networkNodeURL in bc.networkNodes:
        responseAll.append(requests.post(f"{networkNodeURL}/transaction",json=newTransaction).status_code)
    if responseAll.count(200) == len(responseAll):
        return {"note":"Transaction created and broadcast successfully. "}
    else:
        return{"note":"Errorr"}

@app.route('/mine',methods=['GET'])
def mine():
    lastBlock = bc.getLastBlock()
    previousBlockHash = lastBlock['hash']
    currentBlockData = {
        "transactions":bc.pendingTransactions,
        "index":lastBlock['index']+1
    }
    nonce = bc.prooOfWork(previousBlockHash,currentBlockData)
    blockHash = bc.hashBlock(previousBlockHash,currentBlockData,nonce)
    newBlock = bc.createNewBlock(nonce,previousBlockHash,blockHash)
    responseAll = []
    for networkNodeURL in bc.networkNodes:
        res = requests.post(f"{networkNodeURL}/receive-new-block",json={"newBlock":newBlock})
        responseAll.append(res.status_code)
    # for networkNodeURL in bc.networkNodes:
    #     responseAll.append(request.post(f"{networkNodeURL}/transaction/broadcast",json={"newBlock":newBlock}).status_code)
    if responseAll.count(200) == len(responseAll):
        return {
            "note":"Transaction created and broadcast successfully. ",
            "block":newBlock
        }
    else:
        return{"note":"Errorr"}

@app.route("/receive-new-block",methods=['POST'])
def receive_block():
    newBlock = request.json['newBlock']
    lastBlock = bc.getLastBlock()
    correctHash = lastBlock['hash'] == newBlock['previousBlockHash']
    correctIndex = lastBlock['index']+1 == newBlock['index']
    if correctHash and correctIndex:
        bc.chain.append(newBlock)
        bc.pendingTransactions = []
        return {
            "note":"Transaction created and broadcast successfully. ",
            "newBlock":newBlock
        }
    else:
        return {
            "note":"New Block rejected",
            "newBlock":newBlock
        }


@app.route("/register-and-broadcast-node",methods=['POST'])
def register_and_broadcast_node():
    newNodeURL = request.json['newNodeURL']
    try:
        bc.networkNodes.index(newNodeURL)
    except:
        bc.networkNodes.append(newNodeURL)
    responseAll = []
    for networkNodeURL in bc.networkNodes:
        responseAll.append(requests.post(f"{networkNodeURL}/register-node",json={"newNodeURL":newNodeURL}).status_code)
    if responseAll.count(200) == len(responseAll):
        requests.post(f"{networkNodeURL}/register-nodes-bulk",json={"allNetworkNodes":bc.networkNodes + [bc.currentNodeURL]})
        return {"note":"New node registered with network successfully. "}
    else:
        return {"note":"New node registered with network rejected. "}

@app.route("/register-node",methods=['POST'])
def register_node():
    newNodeURL = request.json['newNodeURL']
    try:
        bc.networkNodes.index(newNodeURL)
    except:
        nodeNotAlreadyPresent = True
    notCurrentNode = bc.currentNodeURL != newNodeURL
    if nodeNotAlreadyPresent and notCurrentNode:
        bc.networkNodes.append(newNodeURL)
    return {"note":"New node registered successfully with node. "}
    # return "Register Node"

@app.route("/register-nodes-bulk",methods=['POST'])
def register_nodes_bulk():
    allNetworkNodes = request.json['allNetworkNodes']
    for networkNodeURL in allNetworkNodes:
        try:
            bc.networkNodes.index(networkNodeURL)
        except:
            nodeNotAlreadyPresent = True
        notCurrentNode = bc.currentNodeURL != networkNodeURL
        if nodeNotAlreadyPresent and notCurrentNode:
            bc.networkNodes.append(networkNodeURL)
    return {'note':'Bulk registration successful. '}

@app.route("/consensus",methods=['POST'])
def consensus():
    newLongestChain = []            
    newPendingTransactions = None
    for networkNodeURL in bc.networkNodes:
        # responseAll.append(request.get(f"{networkNodeURL}/blockchain").status_code)
        res = requests.get(f"{networkNodeURL}/blockchain")
        blockchain = res.json()
        if res.status_code == 200:    
            currentChainLength = len(bc.chain)
            maxChainLength = currentChainLength
            if len(blockchain["chain"]) > maxChainLength:
                maxChainLength = len(blockchain["chain"])
                newLongestChain = blockchain["chain"]   
                newPendingTransactions = blockchain["pendingTransactions"]         
    if newLongestChain == [] or (newLongestChain and not bc.chainIsValid(newLongestChain)):
        return {"note":"Current chain has not been replaced. ","chain":bc.chain}
    elif newLongestChain and bc.chainIsValid(newLongestChain):
        bc.chain = newLongestChain
        bc.pendingTransactions = newPendingTransactions
        return {
            "note":"This chain has been replaced. ",
            "chain":bc.chain
        }


@app.route("/block/<string:blockHash>",methods=['GET'])
def block_hash(blockHash):
    blockHash = request.json()['blockHash']
    correctBlock = bc.getBlock(blockHash)
    return {
        "block":correctBlock
    }
@app.route("/transaction/<int:transactionId>",methods=['GET'])
def getTransactionId(transactionId):
    transactionData = bc.getTransaction(transactionId)
    return {
        "transaction":transactionData['transaction'],
        "block":transactionData["block"]
    }

@app.route("/address/<string:address>",methods=['GET'])
def getAddress(address):
    addressData = bc.getAddressData(address)
    return {
        "addressData":addressData
    }

@app.route("/block-explorer",methods=['GET'])
def blockExplorer():
    return render_template('index.html')



if __name__ == '__main__':
    app.run(debug=True,port=args.port)